### PR TITLE
Some systemd fixes revolving dbus and socket activation of systemd services

### LINKED ISSUE #85855: systemd: ship dbus system services
**Describe the bug**
Systemd ships dbus system services (in `${pkgs.systemd}/share/dbus-1/system-services/*.service`), which are used for dbus system activation (see https://dbus.freedesktop.org/doc/system-activation.txt)

This includes some of the units shipped with systemd itself:
```
-r--r--r-- 1 root root 482 Jan  1  1970 org.freedesktop.hostname1.service
-r--r--r-- 1 root root 513 Jan  1  1970 org.freedesktop.import1.service
-r--r--r-- 1 root root 513 Jan  1  1970 org.freedesktop.locale1.service
-r--r--r-- 1 root root 511 Jan  1  1970 org.freedesktop.login1.service
-r--r--r-- 1 root root 515 Jan  1  1970 org.freedesktop.machine1.service
-r--r--r-- 1 root root 515 Jan  1  1970 org.freedesktop.network1.service
-r--r--r-- 1 root root 515 Jan  1  1970 org.freedesktop.resolve1.service
-r--r--r-- 1 root root 462 Jan  1  1970 org.freedesktop.systemd1.service
-r--r--r-- 1 root root 517 Jan  1  1970 org.freedesktop.timedate1.service
-r--r--r-- 1 root root 517 Jan  1  1970 org.freedesktop.timesync1.service
```

We currently don't have the proper tooling ready to make these available in NixOS.

Also, we shouldn't blindly enable all. For example, we might not want to enable `share/dbus-1/system-services/org.freedesktop.resolve1.service, if resolved is enabled.

Originally discovered in #85794, but it always hasn't worked, so it's not a regression.

**Notify maintainers**
@andir @NinjaTrappeur @Mic92 @arianvp

### PR DESCRIPTION
**Pull request by arianvp**
###### Motivation for this change

Resolved wasnt being activated by dbus; which caused DNS errors to occur shortly after or during `nixos-rebuild switch`'ing  . I took the time to fix the activation for all upstream systemd services.

Also fixed socket activation for `systemd-networkd`. so that it can be activated by netlink messages.

Second try of https://github.com/NixOS/nixpkgs/pull/87263

Fixes https://github.com/NixOS/nixpkgs/issues/85855

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by arianvp** [comment]
@GrahamcOfBorg test systemd-timesyncd systemd-networkd systemd-networkd-vrf  networking.networkd systemd

**Comment by arianvp** [comment]
I'm not sure why the networkd-vrf test failed on `aarch64`. seems flakey to me. let me try again:

@GrahamcOfBorg test systemd-timesyncd systemd-networkd systemd-networkd-vrf networking.networkd systemd

File: nixos/modules/system/boot/networkd.nix [hunk 1]
```diff
@@ -1178,14 +1178,19 @@ in
       users.users.systemd-network.group = "systemd-network";
 
       systemd.additionalUpstreamSystemUnits = [
-        "systemd-networkd.service" "systemd-networkd-wait-online.service"
+        "systemd-networkd.socket" "systemd-networkd.service" "systemd-networkd-wait-online.service"
```
Comment (flokli) [inline nixos/modules/system/boot/networkd.nix:1181]: please put these in a new line each, and sort alphabetically.

File: nixos/modules/system/boot/networkd.nix [hunk 2]
```diff
@@ -1178,14 +1178,19 @@ in
...
       ];
 
       systemd.network.units = mapAttrs' (n: v: nameValuePair "${n}.netdev" (netdevToUnit n v)) cfg.netdevs
         // mapAttrs' (n: v: nameValuePair "${n}.network" (networkToUnit n v)) cfg.networks;
 
+      # systemd-networkd is socket-activated by kernel netlink route change
+      # messages. Important to have systemd buffer those on behalf of networkd.
```
Comment (flokli) [inline nixos/modules/system/boot/networkd.nix:1188]:
```suggestion
      # messages. It is important to have systemd buffer those on behalf of networkd.
```

**Review by flokli — APPROVED** [review]
LGTM otherwise (but haven't tested myself so far)

**Comment by flokli** [comment]
Closing in favour of #89064.

**Comment by flokli** [comment]
Reopened as per https://github.com/NixOS/nixpkgs/pull/89064#issuecomment-636753727.

**Review by matthewbauer — APPROVED** [review]
Looks good to me - it appears to fix the network loss on nixos-rebuild.

I'll let someone else with a little more knowledge on systemd merge though.

Probably should be backported to nixos-20.03.

**Comment by flokli** [comment]
@arianvp can you address the changes requested in this PR?

As https://github.com/NixOS/nixpkgs/pull/89064#issuecomment-636756210 hasn't received any feedback, I propose merging this one first, and then doing #89064 later.

**Comment by flokli** [comment]
@matthewbauer This has been broken for quite some time, it's not a regression with 20.03.

If we still merge this to 20.03, IMHO it should get at least a month of testing in nixos-unstable before thinking about backporting.

**Comment by matthewbauer** [comment]
> @matthewbauer This has been broken for quite some time, it's not a regression with 20.03.
> 
> If we still merge this to 20.03, IMHO it should get at least a month of testing in nixos-unstable before thinking about backporting.

Hmmm... yeah it definitely was a regression that I started noticing since 19.03 or 19.09. It doesn't always happen, so maybe I shouldn't be so sure this actually fixed it. It seems like a rebuild that touches network configuration options make it more likely.

File: nixos/modules/system/boot/networkd.nix (hunk 1 above)
Comment (flokli) [reply nixos/modules/system/boot/networkd.nix]: I addressed this by myself.

File: nixos/modules/system/boot/networkd.nix (hunk 2 above)
Comment (flokli) [reply nixos/modules/system/boot/networkd.nix]: I addressed this by myself.

**Comment by flokli** [comment]
I addressed the required changes and will merge this in now.
